---
title: "Fusariose: QTLs Description - Part1"
author: "Yan Holtz"
date: "October 2016"
output:
  rmdformats::readthedown:
    highlight: kate
  html_document:
    toc: yes
    toc_depth: 3
---



[//]: =========================================================================================================================
#1/ Introduction
This file aims to describe the QTLs found for fusariose resistance.   
It concerns the evaluation of more than 300 variables, so it can be a bit long to run it.  
Let's upload the file containing all the LOD scores. It contains about 4M lines, so it is quite long...(~2minutes)
```{r}
#Watch out, to reproduct analysis, you have to update the path.
data=read.table("~/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/bilan_simple_marker.gz" , header=T , sep="," )
data=data[ order(data$LG , data$Distance) , ]
# random lines of the dataset to develop quicker
#data=data[ sample( seq(1,nrow(data)) , 30000 )  , ]
#data=data[ order(data$LG , data$Distance) , ]
```

Let's upload the physical position of every genes of the BWr
```{r}
POS=read.table("/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/physical_map_of_BW.txt", na.string="-" , header=T , dec=".") ; colnames(POS)=c( "chromo_BW" ,  "contig" ,"position_BW" )
```

Let's upload the genotyping matrix:
```{r}
# Genotype matrix
geno = read.table(file="~/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/fichier_genotypage_QTL.csv", sep = ";" , header = F, na.strings = "-")
geno= as.matrix(geno)
colnames(geno)= geno[1,]
geno= as.data.frame(geno[-1 , ])
# Remove useless individuals
#geno <- geno [ geno[,1] %in% names(data),]
```

Let's upload the phenotyping matrix:
```{r}
#Watch out, to reproduct analysis, you have to update the path.
pheno=read.table("~/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/PHENOTYPE/phenotypage_all_fusa.csv" , header=T , sep=";" )
#numeric
pheno[,-1]=apply(pheno[,-1],2,as.numeric)
#check no redondancy in genotype name
table(pheno$geno)[table(pheno$geno)>1]
# put geno name as rowname
rownames(pheno)=pheno$geno
pheno=pheno[,-1]
# delete columns with only NA
which(apply( pheno , 2 , function(x) all(is.na(x)) )==TRUE)
pheno=pheno[ , ! apply( pheno , 2 , function(x) all(is.na(x)) ) ]
```


Charge some libraries that will be useful
```{r results='hide', message=FALSE, warning=FALSE}
library(RColorBrewer)
library(xtable)
library(plotly)
library(FactoMineR)
```

Let's create a color vector for the study
```{r}
my_colors = brewer.pal(8, "Set2") 
my_colors = colorRampPalette(my_colors)(20)
```

How many variables do we have? --> 312 phenotypic variables.
```{r}
my_carac=levels(data$variable)
nlevels(data$variable)
```

What is the LOD threshold to declare a QTL significant?
```{r}
my_QTL_thres=3.73
```


Let's do some object that contains group of variable:
```{r}
# reps
REP1=my_carac[grep("rep1", my_carac )]
REP2=my_carac[grep("rep2", my_carac )]
BLUP=my_carac[grep("blup", my_carac )]

# experiments
CAP13=my_carac[grep("CAP13", my_carac )]
GRI11=my_carac[grep("GRI11", my_carac )]
GRI13=my_carac[grep("GRI13", my_carac )]
GRI15=my_carac[grep("GRI15", my_carac )]
LEC14=my_carac[grep("LEC14", my_carac )]
BEQ11=my_carac[grep("BEQ11", my_carac )]
BAR14=my_carac[grep("BAR14", my_carac )]

# Type de phéno NEPI, NEPIL, PEPIL, NOT, DON
PEPIL=my_carac[grep("PEPIL", my_carac )]
PEPI=my_carac[grep("PEPI", my_carac)]
PEPI=setdiff(PEPI, PEPIL)
NEPIL=my_carac[grep("NEPIL", my_carac )]
NOT=my_carac[grep("NOT", my_carac )]
DON=my_carac[grep("DON", my_carac )]

# AUDPC et derniere notation
AUDPC=my_carac[grep("AUDPC", my_carac )]
LAST=c("CAP13_PEPI550_rep1","CAP13_NEPIL550_rep1","CAP13_PEPIL550_rep1","CAP13_PEPI550_rep2","CAP13_NEPIL550_rep2","CAP13_PEPIL550_rep2","CAP13_PEPI550_blup","CAP13_NEPIL550_blup","CAP13_PEPIL550_blup","GRI11_NOT550_rep1","GRI11_NEPIL550_rep1","GRI11_NOT550_rep2","GRI11_NEPIL550_rep2","GRI13_PEPI550_rep1","GRI13_NEPIL550_rep1","GRI13_PEPIL550_rep1","GRI13_NOT550_rep1","GRI13_PEPI550_rep2","GRI13_NEPIL550_rep2","GRI13_PEPIL550_rep2","GRI13_NOT550_rep2","GRI13_PEPI550_blup","GRI13_NEPIL550_blup","GRI13_PEPIL550_blup","GRI13_NOT550_blup","LEC14_PEPI550_rep1","LEC14_NEPIL550_rep1","LEC14_PEPIL550_rep1","LEC14_PEPI550_rep2","LEC14_NEPIL550_rep2","LEC14_PEPIL550_rep2","LEC14_PEPI550_blup","LEC14_NEPIL550_blup","LEC14_PEPIL550_blup","BEQ11_PEPI450","BEQ11_NOT500","BEQ11_NEPIL500","BAR14_PEPI600_rep1","BAR14_NEPIL600_rep1","BAR14_PEPIL600_rep1","BAR14_PEPI600_rep2","BAR14_NEPIL600_rep2","BAR14_PEPIL600_rep2","BAR14_PEPI600_blup","BAR14_NEPIL600_blup","BAR14_PEPIL600_blup")

# agronomical variable
EPI=my_carac[grep("_EPI" , my_carac )]
FLO=my_carac[grep("FLO" , my_carac )]
HEI=my_carac[grep("HEI" , my_carac )]
VER=my_carac[grep("VER" , my_carac )]

# Resistance set for publication = BLUps when avail. DO we have all the needed traits?
a=c(DON, NOT, PEPI, NEPIL, PEPIL)
b=intersect(a,BLUP)
c=intersect(BEQ11, a)
PUBLI=c(b,c,"LEC14_DON")
```


Let's check what is available for which experiment?
```{r}
# start table
bilan=data.frame( matrix(0,5,7) )
colnames(bilan)=c("CAP13", "GRI11", "GRI13", "GRI15", "LEC14", "BEQ11","BAR14")
rownames(bilan)=c("DON?", "PEPI?", "NOT?", "NEPIL?", "PEPIL?")
all=list(CAP13, GRI11, GRI13, GRI15, LEC14, BEQ11, BAR14)
num1=0
num2=0
for(i in all){
  num1=num1+1
  num2=0
  for(j in list(DON, PEPI, NOT, NEPIL, PEPIL)){
   num2=num2+1
   a=intersect(i, j)
   bilan[ num2 , num1]=length(a)
  }}
```

```{r my_tabledkokd1, results='asis' , echo=FALSE}
print(xtable(bilan), type = "html", include.rownames = T , comment=FALSE)
```

We know that:  
- DON is available for CAP13, GRI13 and BAR14 only.
- PEPI is not available for GRI11
- NOT is not available for CAP13.
- PEPIL is not available for GRI11 and BEQ11.
So please check it is indeed what we have in this QTL file!

[//]: =========================================================================================================================












[//]: =========================================================================================================================
#2/ Useful functions

```{r}
#A function to plot a QTL and visualize the LOD scores along the chromosome. Works for several variable on a specific chromosome.
plot_a_QTL=function(data, select_chromo, variable, LOD_threshold, my_ylim){
  
  # order data
  data=data[ order(data$LG , data$Distance) , ]

  # plot the first variable
  a=data[data$variable==variable[1] & data$LG==select_chromo & !is.na(data$LOD) , ]
  plot(a$LOD ~ a$Distance , type="l" , ylim=c(0,my_ylim) , ylab="LOD score" , xlab="position (cM)" , col=my_colors[1] , lwd=1.6 )
  
  # for every other variable, add a LOD score
  num=1
  for( i in c(2:length(variable))){
    num=num+1
    a=data[data$variable==variable[num] & data$LG==select_chromo & !is.na(data$LOD) , ]
    points(a$LOD ~ a$Distance , type="l" , ylim=c(0,my_ylim) , ylab="LOD score" , xaxt="n" , col=my_colors[num] , lwd=1.6 )
    }
  
  # add lod threshold
  abline(h=my_QTL_thres , col="grey")

  # add legend
  legend("topright" , horiz=F , col=my_colors[c(1:length(variable))] , legend=variable , bty="n" , lty=1 , pt.cex=2, lwd=1.6)
  legend("topleft" , horiz=F , col="white" , legend=select_chromo , bty="n" , lty=1 , pt.cex=0, lwd=1.6)

  }

#Same function but with interactive graphics!
plot_a_interactiveQTL=function(data, select_chromo, variable, LOD_threshold, my_ylim){

# keep selected data & order it
a=data[data$variable%in%variable & data$LG==select_chromo & !is.na(data$LOD) , ]
a=a[ order(a$LG , a$Distance) , ]

#When I have several markers at the same position, I replace the LOD-scores by the highest LOD-scores of the position
for(i in nrow(a)){
  highest=max(a$LOD[which(a$variable==a$variable[i] & a$Distance==a$Distance[i])])  
  a$LOD[i]=highest
}

#create text for hover
a$text=paste(a$variable,"<br>",a$marqueur,"<br>","allèles A et B: ",round(a$moy.A,2)," | ",round(a$moy.B,2), "<br>","Pos gen: ", a$Distance , "<br>", "Pos phy: ", a$group_physique , " | ", a$Posi_physique , sep="")

# plot
plot_ly( x=a$Distance, y=a$LOD, type="scatter", mode="lines", text=a$text, hoverinfo="text", color=a$variable)%>%
add_trace(x=seq.int(from=0, to=max(a$Distance,na.rm=T), length.out = nrow(a) ), y=rep(LOD_threshold, nrow(a)), type="scatter", mode = "lines", line=list(color="grey"), hoverinfo="none") %>%
layout(
  showlegend=F,
  hovermode="closest",
  xaxis=list(title="Genetic position (cM)"),
  yaxis=list(title="LOD score" , range=c(0, my_ylim) )
  )
}

#A function to summarize all the significant QTL of a list of variables. Gives the max LOD score, the IC, the position etc..
summary_table_for_QTL=function(data, selected_variable, LOD_threshold, size_IC){

  # order data
  data=data[ order(data$LG , data$Distance) , ]
  my_colors=brewer.pal(5,"Paired")
  
  # create empty summary table
  bil=data.frame(matrix(0,1000,10))
  colnames(bil)=c("pop","carac","chromo","LOD_max","position","marker","IC","R2","a","diff")
  
  # loop to study every variable & chromosome
  num_line_bilan=0
	num_chromo=0

	for(chrom in levels(data$LG)){
		num_variable=0
		num_col=0
		num_chromo=num_chromo+1

		for(var in selected_variable){
			num_col=num_col+1
			current_data=data[ data$variable==var & data$LG==chrom  , ]
			current_data=current_data[!is.na(current_data$LOD) , ]
			signif_data=current_data[ current_data$LOD > LOD_threshold  , ]
			
			#Si j'ai des marqueurs significatifs
			if(nrow(signif_data)>0){
				
				#Je récupère les infos de ce QTL
				LOD_max=max(signif_data$LOD)
				mark_max=signif_data$marqueur[signif_data$LOD==LOD_max]
				pos_max=signif_data$Distance[signif_data$LOD==LOD_max]
				r2_max=signif_data$R2[signif_data$LOD==LOD_max]
				a_max=signif_data$a[signif_data$LOD==LOD_max]
	      my_diff=ifelse( signif_data$moy.A[signif_data$LOD==LOD_max][1] > signif_data$moy.B[signif_data$LOD==LOD_max][1] , "Dic2>Silur" , "Silur>Dic2" )
	        
				#Détermination IC --> je bloque la zone max a 30 cM
				in_IC=current_data[ current_data$LOD > (LOD_max-size_IC) & current_data$Distance > (pos_max-30) & current_data$Distance < (pos_max+30)  , ]
				IC_min=min(round(in_IC$Distance,2))
				IC_max=max(round(in_IC$Distance,2))
				
				#Je remplie le tableau bilan
				num_line_bilan=num_line_bilan+1
				bil[num_line_bilan , 1]="DS"
				bil[num_line_bilan , 2]=var
				bil[num_line_bilan , 3]=chrom
				bil[num_line_bilan , 4]=LOD_max
				bil[num_line_bilan , 5]=pos_max[1]
				bil[num_line_bilan , 6]=as.character(mark_max)[1]
				bil[num_line_bilan , 7]=paste(IC_min, IC_max, sep="-")
				bil[num_line_bilan , 8]=r2_max[1]
				bil[num_line_bilan , 9]=a_max[1]
				bil[num_line_bilan , 10]=my_diff
				
				#close 3 loops
				}}}
	
# Clean and print the Table
bil=bil[bil$pop!=0 , ]
bil=bil[ order(bil$pop , bil$chromo , bil$carac) , ]
return(bil)
#close function
}

#A function to represent confidence interval of QTLs on the genetic map.
show_IC_QTL_on_map=function(data, selected_variable, selected_chromosome, LOD_threshold, size_IC){

  # user can pick every variable
  if(selected_variable[1]=="all"){selected_variable=levels(data$variable)}
  
  # order data
  data=data[ order(data$LG , data$Distance) , ]
  my_colors=brewer.pal(12,"Set3")
  my_colors = colorRampPalette(my_colors)(30)
  var_with_QTL=c()
  
  # loop to study every variable & chromosome
	num_chromo=0

	# prepare background of chart 
	nchro=length(selected_chromosome)
	xchro=c(seq(5,5*nchro,5) )
  par(mar=c(0,5,2,1))
  plot(1,1,col="transparent",bty="n", xaxt="n", yaxt="l" , xlab="", ylab="position (cM)" , xlim=c(3,max(xchro)+11) , ylim=c(320,-5) )
  num=0
  for(chrom in selected_chromosome){
  	num=num+1
  	A=data[data$variable==selected_variable[1] & data$LG==chrom, ]
  	points(rep(xchro[num],nrow(A)) , A$Distance , pch=20 , cex=0.8 , col="grey")
  	}

	# IC calculation and plot
  for(chrom in selected_chromosome){

		num_variable=0
		num_chromo=num_chromo+1
		num_col=0

		for(var in selected_variable){
		  num_col=num_col+1
			current_data=data[ data$variable==var & data$LG==chrom  , ]
			current_data=current_data[!is.na(current_data$LOD) , ]
			signif_data=current_data[current_data$LOD>LOD_threshold  , ]
			
			#Si j'ai des marqueurs significatifs
			if(nrow(signif_data)>0){
			  
			  # I add it to the list of variables with QTL
			  if(!var%in%var_with_QTL){var_with_QTL=c(var_with_QTL,var)}
				
				# Je récupère les infos de ce QTL
				LOD_max=max(signif_data$LOD)
				mark_max=signif_data$marqueur[signif_data$LOD==LOD_max]
				pos_max=signif_data$Distance[signif_data$LOD==LOD_max]

				# Détermination IC --> je bloque la zone max a 30 cM
				in_IC=current_data[ current_data$LOD > (LOD_max-size_IC) & current_data$Distance > (pos_max-30) & current_data$Distance < (pos_max+30)  , ]
				IC_min=min(in_IC$Distance)
				IC_max=max(in_IC$Distance)
				
				# J'ajoute le trait corresponant a ma variable
				num_variable=num_variable+0.2
				lines(c(xchro[num_chromo]+num_variable,xchro[num_chromo]+num_variable) , c(IC_min,IC_max) , col=my_colors[which(var_with_QTL%in%var)], lwd=6)
				num_variable=abs(num_variable)

        #Close 2 loops and a if
				}}}
	
#Ajout légende et nom de chromosome?
text(xchro , rep(-10,3) , selected_chromosome , col="orange")
#legend("bottomright", 14 , 240 , horiz=F , col=my_colors[1:length(var_with_QTL)] , legend=var_with_QTL , bty="n" , lty=1 , lwd=6 )

#Close function
}

#A function that gives IC features: size, # of markers, # of genes...
find_IC_infos=function(min_IC, max_IC, chrom, to_remove){

  # Get the part of the genetic map concerned by this IC:
  IC_map=data[ which(data$LG==chrom & data$Distance>=min_IC & data$Distance<=max_IC & data$LG==data$group_physique & data$variable==levels(data$variable)[1]), c(1:5)]

  # remove markers if needed.
  if(length(to_remove)>0){
    IC_map=IC_map[-which(IC_map$marqueur%in%to_remove) , ]
  }
  
  # Calculate statistics

  # genet
  deb_IC_gen=min(IC_map$Distance)
  end_IC_gen=max(IC_map$Distance)
  size_IC_gen=end_IC_gen - deb_IC_gen
  size_IC_gen=round(size_IC_gen , 2)
  nb_of_mark=nrow(IC_map)

  # physic
  deb_IC_phy=min(IC_map$Posi_physique)
  end_IC_phy=max(IC_map$Posi_physique)
  size_IC_phy=end_IC_phy - deb_IC_phy
  nb_of_gene=nrow(POS[which(POS$group==chrom & (POS$position*1000000)>=deb_IC_phy & (POS$position*1000000)<=end_IC_phy) , ])

  # return a vec with all info
  vec=c(chrom, min_IC, max_IC, size_IC_gen, nb_of_mark, deb_IC_phy, end_IC_phy, size_IC_phy, nb_of_gene)
  return(vec)
  }

#A function to show all the genes present in a IC. The relatioship between physical and genetic position is also available
show_genes_of_IC=function(min_IC, max_IC, chrom, my_lwd, my_cex, to_remove){
  
  # get data for the IC
  IC_map=data[ which(data$LG==chrom & data$Distance>=min_IC & data$Distance<=max_IC & data$group==data$group_physique & data$variable==levels(data$variable)[1] & data$LG==data$group_physique ) , c(1:5)]

  # remove markers if needed.
  if(length(to_remove)>0){
    IC_map=IC_map[-which(IC_map$marqueur%in%to_remove) , ]
  }

  # physical informations
  #deb_IC_phy=min(IC_map$Posi_physique)
  #end_IC_phy=max(IC_map$Posi_physique)
  my_genes_pos=POS[which(POS$chromo_BW==chrom) , ]

  #plot
  plot(IC_map$Distance ~ as.numeric(IC_map$Posi_physique/1000000),
     ylim=c(min_IC, max_IC),
     xlab="physical pos. of markers (Mb)", ylab="genetic position (cM)" , 
     main=paste("chromome ",chrom ," -- IC : ",min_IC," - ",max_IC," cM",sep="" ),
     col=rgb(0.4,0.3,0.6,0.4) , pch=20 , cex=my_cex
     )
  abline(v=my_genes_pos[,3]/1000000 , col="grey", lwd=my_lwd)

  }
```

TODO
A function to do an interactive representation of the IC, with physical and genetic correspondances..
```{r, eval=FALSE}

#### A FINIR A L'OCCAS

#get data for the IC
min_IC=1
max_IC=9
chrom="1B"
IC_map=data[ which(data$LG==chrom & data$Distance>=min_IC & data$Distance<=max_IC & data$group==data$group_physique & data$variable==levels(data$variable)[1]), c(1:5)]

# Make a Y vector for the connection between markers
pos=as.matrix(na.omit(IC_map[,c(3,5)]))
my_vect=as.vector(t(pos))
correctif=seq(1:length(my_vect)) + rep(c(0,0,1,-1) , length.out=length(my_vect) )  
my_vect=my_vect[correctif]
if(length(my_vect)%%4 == 0){ my_vect=c(my_vect , my_vect[length(my_vect)] , my_vect[length(my_vect)-1]) } 
yaxis=my_vect

# Make a X vector
xaxis=rep( c(1,2,2,1), nrow(IC_map)/2 )
if(length(xaxis)%%4 == 0){ xaxis=c(xaxis , xaxis[length(xaxis)] , xaxis[length(xaxis)-1]) } 

# Make the interactive graphic
p=plot_ly(x=xaxis , y=yaxis , hoverinfo="none", line=list(width=1, color="grey" , opacity=0.1) , showlegend=F)    
#layour
p=layout( 
  hovermode="closest"  ,
  xaxis=list(title = "", zeroline = FALSE, showline = FALSE, showticklabels = FALSE, showgrid = FALSE , range=c(0.5,nb_selected_maps+0.5) ),
  yaxis=list(range=c(0,500), autorange = "reversed", title = "Position (cM)", zeroline = F, showline = T, showticklabels = T, showgrid = FALSE   ,  tickfont=list(color="grey", size=15) , titlefont=list(color="grey", size=15) , tickcolor="grey" , linecolor="grey")
)
# line for 2 chromosomes:
p=add_trace( x=c(1,2), y=c(0, max(IC_map$Posi_physique,na.rm=T)) , evaluate=TRUE , line=list(width=4, color="black"),showlegend=F )
p=layout( yaxis=list(range=c(0,IC_map$Posi_physique)) )
p
```


A function to boxplot a phenotype following 2 alleles?
```{r}
# A function to represent a QTL. It makes a boxplot with 2 categories: alleles A and B
boxplot_a_QTL=function(marker, trait, ...){

  # prepare data: merge expression and alleles:
  nucl= geno[,c(1,which(colnames(geno)==marker)) ]
  my_pheno= data.frame(ind=rownames(pheno), pheno[ , trait])
  AA=merge(my_pheno, nucl, by.x=1, by.y=1, all=T)

  # Compute means
  my_means=aggregate(AA[,2] , by=list(AA[,3]), mean, na.rm=T)
  my_meanA=round(my_means[1,2] , 2)
  my_meanB=round(my_means[2,2] , 2)
  # T test
  my_ttest=round(t.test(AA[,2]~AA[,3])$p.value , 2)

  # Basic boxplot
  boxplot(AA[,2]~AA[,3], 
        col=my_colors[c(5,6)], 
        ...,
        ylab=paste("value of ",gsub("\\|T.*","",trait),sep=""), 
        xlab=paste("allele of ",marker,sep="") , 
        main=paste("A: ",my_meanA," | B: ",my_meanB," | pval: ",my_ttest, sep="" ) 
        )

  # Add data points
  mylevels<-levels(AA[,3])
  levelProportions<-summary(AA[,3])/nrow(AA)
  for(i in 1:length(mylevels)){
    thislevel<-mylevels[i]
    thisvalues<-AA[AA[,3]==thislevel, 2]
    # take the x-axis indices and add a jitter, proportional to the N in each level
    myjitter<-jitter(rep(i, length(thisvalues)), amount=levelProportions[i]/2)
    points(myjitter, thisvalues, pch=20, col=rgb(0,0,0,.2)) 
  }
}

#A function to calculate interaction between 2 QTLs for a trait
# Function to calculate interaction effect between QTLs:
analyse_inter=function(marker1, marker2, trait){

  # prepare data: merge expression and alleles:
  nucl= geno[ , c(1, which(colnames(geno)%in%c(marker1,marker2)) ) ]
  my_pheno= data.frame(ind=rownames(pheno), pheno[ , trait])
  AA=merge(my_pheno, nucl, by.x=1, by.y=1, all=T)
	AA=na.omit(AA)

	# Prepare
	my_mark=paste(AA[,3],AA[,4],sep="-")
	means <- round(tapply(AA[,2],my_mark,mean,na.rm=T) ,2)
	
	# Calculate interaction
	model=lm(AA[,2] ~ AA[,3] * AA[,4])
	sum=summary(model)
	tot_r2=round(sum$r.squared,3)
	inter_pval=round(sum$coefficients[4,4],10)
	
	# Complete summary files
	res=c(trait, marker1, marker2, means, tot_r2, inter_pval)
	return(res)
}

#A function to boxplot 2 QTLs together
# A function to represent a QTL. It makes a boxplot with 2 categories: alleles A and B
boxplot_two_QTL=function(marker1, marker2, trait, ...){

  # prepare data: merge expression and alleles:
  nucl= geno[ , c(1, which(colnames(geno)%in%c(marker1,marker2)) ) ]
  my_pheno= data.frame(ind=rownames(pheno), pheno[ , trait])
  AA=merge(my_pheno, nucl, by.x=1, by.y=1, all=T)
	AA=na.omit(AA)

	# Prepare
	my_mark=paste(AA[,3],AA[,4],sep="-")
	means <- round(tapply(AA[,2],my_mark,mean,na.rm=T) ,2)

  #plot
	par( mar=c(5,5,2,2))
  boxplot(AA[,2] ~ my_mark , 
      medlwd=0, cex.axis=0.6, cex.col="grey", las=1, main="" ,
      ...,
      ylab=trait,  xlab="Bi-locus genotype" , col=my_colors[c(3,6,8,3)] , xaxt="n", boxwex=0.4) 
  my_labels=c( expression(paste('Dic2'["1B"],"-",'Dic2'["5A"])) , 
               expression(paste('Dic2'["1B"],"-",'Silur'["5A"])) ,  
               expression(paste('Silur'["1B"],"-",'Dic2'["5A"])) , 
               expression(paste('Silur'["1B"],"-",'Silur'["5A"])) )
  axis(labels=my_labels , at=c(1,2,3,4) , side=T)
}
```



A function to boxplot 2 QTLs together -> INTERACTIVE
```{r}
# A function to represent a QTL. It makes a boxplot with 2 categories: alleles A and B
boxplot_two_QTL_interactive=function(marker1, marker2, trait, ...){

  # example to develop if needed
  #marker1="Cluster_16778|Contig1|original@1840"
  #marker2="Cluster_9940|Contig1|complementarySeq@104" 
  #trait="BAR14_PEPIL300_blup"

  # prepare data: merge expression and alleles + add dic2 and Silur
  nucl= geno[ , c(1, which(colnames(geno)%in%c(marker1,marker2)) ) ]
  qq=data.frame(c("dic2","silur"), c("A","B"), c("A","B"))
  colnames(qq)=colnames(nucl)
  nucl=rbind(nucl,qq)
  my_pheno= data.frame(ind=rownames(pheno), pheno=pheno[ , trait])
  AA=merge(my_pheno, nucl, by.x=1, by.y=1, all=T)
	AA=na.omit(AA)

	# Prepare
	AA$my_mark=as.factor(paste(AA[,3],AA[,4],sep="-"))
	means <- round(tapply(AA[,2],AA$my_mark,mean,na.rm=T) ,2)
  AA=AA[order(AA$my_mark),]
  AA$text=paste("ind: ",AA$ind,"<br>","value: ",round(AA$pheno,2),sep="")

  # First we calculate a new x axis with the jiiter added to it. Jitter level must be relative to the number of points per level.
  my_prop=summary(AA$my_mark) / nrow(AA)
  new_x=c()
  for(i in c(1:nlevels(AA$my_mark))){
    myjitter<-jitter( rep(i, summary(AA$my_mark)[i]), amount=my_prop[i]/2)
    new_x=c(new_x, myjitter)
    }

# Plot
plot_ly(AA) %>%
	add_trace( x=as.numeric(AA$my_mark), y = ~pheno, color = ~my_mark, type = "box" ) %>%
	add_trace( x=new_x, y = ~AA$pheno,  
		type = "scatter", 
		mode="markers",  
		marker = list(size = 10, color = ifelse(AA$ind=="dic2","blue",ifelse(AA$ind=="silur","red",rgb(0.1, 0.1, 0.1, 0.3))), line = list(color = rgb(0.1, 0.1, 0.1, 0.8), width = 2)),
		text=~text,
		hoverinfo="text"
	)%>%
	layout(title = "",
	       hovermode="closest",
         showlegend=F,
         xaxis = list(title = "" , tickfont = list( color="orange", size=26 ),  tickmode="array", tickvals=c(1,2,3,4) , ticktext=levels(AA$my_mark) , showline=T  ),
         yaxis = list (title = colnames(data)[2] , gridwidth=2, zeroline=F)
         
    )

}
```

[//]: =========================================================================================================================








[//]: =========================================================================================================================
#3/ Determine LOD threshold
We used R-QTL to do some permutation and determine which is the LOD threshold over which q QTL can be considered as significant.
(This is done on a cluster because of long computation time).
A lod threshold was calculated for each trait.
The highest treshold observed was 3.73. We thus used this threshold for every trait since it is the most conservative one.

```
cd ~/work/FUSA/LOD_SEUIL
cd /NAS/g2pop/HOLTZ_YAN_DATA/DIC2_SILUR/QTL/PUBLI
# je créé un script avec:
for i in $(seq 2 318) ; do echo $i ; Rscript /NAS/g2pop/HOLTZ_YAN_DATA/programmes/Calculate_LOD_threshold.R genotypage.csv phenotypage.csv carte $i ; done
# puis j'envoie
qsub -q normal.q -b yes -cwd -N LOD_thres "./script"
more LOD_thres.o* | grep -B1 -A2 "1000 permutations)" | grep -v "1000 perm" | grep -v "lod" | sed 's/\[1\] \"//' | sed 's/\"//' | sed 's/5% //' > LOD_threshold_per_trait
cat LOD_threshold_per_trait | grep -v "-" | grep -v "[A-Z]" > tmp
R
data=read.table("tmp")
max(data[,1])
min(data[,1])
hist(data[,1])
```

[//]: =========================================================================================================================








[//]: =========================================================================================================================
#4/ Agronomical features
Let's check if we found the expected QTLs for height and flowering date (already known in the litterature..)



## height
Supposed to be on chromosome 4B. As you can see on the graphic below, there is no doubt concerning this QTL. It also proove that scripts are working properly. Note you can **zoom** and **hover** the graphics to get more details!
```{r, warning=F}
plot_a_interactiveQTL(data, "4B", HEI  , my_QTL_thres, 40)
```

Same result in a table:
```{r, warning=FALSE}
summary_QTL_HEI=summary_table_for_QTL(data, HEI, my_QTL_thres, 1.5)
```

```{r my_tablzzzfdfvd1, results='asis' , echo=FALSE}
print(xtable(summary_QTL_HEI), type = "html", include.rownames = F , comment=FALSE)
```

We can see that this QTL is really consistent from one experiment to another. Dic2 is taller than Silur. Let's visualize it in a boxplot:
```{r}
boxplot_a_QTL("Traes_4BS_684D5DDDB@884", "CAP13_HEI_rep1"  )
```



## flowering (heading?) date
Flowering date and precocity are known to be controlled by a QTL on chromosome 7B. We also find this result:
```{r}
p1=plot_a_interactiveQTL(data, "7B", EPI  , my_QTL_thres, 30)
p2=plot_a_interactiveQTL(data, "7B", FLO  , my_QTL_thres, 30)
subplot(p1, p2)
```

## Verse?
There is only one QTL for the "verse", on the same position that the QTL that controls height. This is quite logical: the higher is the plant, the more chance it has to fall...
Verse is available for CAP13 only. It is funny to see that the QTL is significant for rep1 and rep2, but not for the blup...
```{r}
plot_a_interactiveQTL(data, "4B", VER , my_QTL_thres, 30)
```



[//]: =========================================================================================================================









[//]: =========================================================================================================================
#5/ Resistance to fusariose





## Toxine concentration (DON)
Let's summarize in a table every significant QTL for the "DON" variable:

```{r, warning=FALSE}
summary_QTL_DON=summary_table_for_QTL(data, DON, my_QTL_thres, 1.5)
```

```{r my_tabledvfdfvd1, results='asis' , echo=FALSE}
print(xtable(summary_QTL_DON), type = "html", include.rownames = F , comment=FALSE)
```

If we consider only blups, there is only one QTL on chromosome 3A, with a LOD of 4.19 for DON in CAP13.  
Let's show this 3A chromosome with the DON variables:
```{r}
plot_a_interactiveQTL(data, "3A", intersect(DON,PUBLI), my_QTL_thres, 6)
```
There is absolutely no effect for other trials than CAP13...






## Type 1: Note FUSA
Summary of all the significant QTLs for the NOT: 

```{r, warning=FALSE}
summary_QTL_NOT=summary_table_for_QTL(data, NOT, my_QTL_thres, 1.5)
```

```{r my_tabledvd2, results='asis', echo=FALSE}
print(xtable(summary_QTL_NOT), type = "html", include.rownames = F , comment=FALSE)
```


I have 2 main QTLs: on the 1B and on the 5A.  
Then I have some really small messages on 2A - 2B - 3A  
And also the QTL linked with height and precocity on chromo 4B and 7B.  
But 4B-height QTL does not appear for BLUPs --> is it due to the fact thatI added height in the model while calculating blups?
  
Let's show the QTLs on graphics. We show only the last note and audpc for blup when available.

```{r}
toshow=c("GRI11_NOT550_blup","GRI11_NOTAUDPC_blup","GRI13_NOT550_blup","GRI13_NOTAUDPC_blup","GRI15_NOT550_blup","GRI15_NOTAUDPC_blup","BEQ11_NOT500","BEQ11_NOTAUDPC","BAR14_NOT600_blup")
#Chromosome 1B
plot_a_interactiveQTL(data, "1B", intersect(NOT,PUBLI), my_QTL_thres, 10)
#Chromosome 5A
plot_a_interactiveQTL(data, "5A", intersect(NOT,PUBLI), my_QTL_thres, 10)
```

Let's visualize an example trough a boxplot:
```{r}
par(mfrow=c(1,2))
boxplot_a_QTL("Traes_1BS_B9CBEFAE5@3346", "GRI15_NOTAUDPC_rep2" )
boxplot_a_QTL("Traes_5AL_A714BA22E@655", "GRI15_NOTAUDPC_rep2")
```






## Type 1: % Spike with Fusa
A lot of significant QTLs.

```{r, warning=FALSE}
summary_QTL_PEPI=summary_table_for_QTL(data, PEPI, my_QTL_thres, 1.5)
```

```{r my_tablevfdbgdfbgfd3, results='asis', echo=FALSE}
print(xtable(summary_QTL_PEPI), type = "html", include.rownames = F , comment=FALSE)
```

Two main QTLs on chromosome 1B and 5A. For the 1B QTLs, it is interesting to see that once more, the resistance allele comes from Silur.. How is it possible? The 1B QTLs works for BAR14 only, but for every traits and with strong LOD-scores.The 5A works for 3 experiments.  
Then we have some small signals on chromosomes 2A, 2B, 3A, 4A, 6A. Nothing really interesting.  
And once more the QTL linked with height and precocity on chromosomes 4B and 7B. For the 4B--> no BLUP.


We can have a look to the LOD-scores of the main traits (AUDPC and LAST):
```{r}
# Variable to show on the plot?
my_var_to_show=c( Reduce(intersect, list(PEPI,BLUP,AUDPC)), Reduce(intersect, list(PEPI,BLUP,LAST))  )
par(mfrow=c(2,2))
#Chromosome 1B
plot_a_interactiveQTL(data, "1B", my_var_to_show, my_QTL_thres, 12)
#Chromosome 5A
plot_a_interactiveQTL(data, "5A", my_var_to_show, my_QTL_thres, 7)
```

An example for QTL 1B and 5A on a boxplot:

Let's visualize an example trough a boxplot:
```{r}
par(mfrow=c(1,2))
boxplot_a_QTL("Traes_1BS_B9CBEFAE5@3346", "BAR14_PEPI600_rep1")
boxplot_a_QTL("Traes_5AL_8682BB101@185", "GRI15_PEPI550_rep2")
```





## Type 2: % of Spikelet with Fusa
A lot of significant QTLs.

```{r, warning=FALSE}
summary_QTL_PEPIL=summary_table_for_QTL(data, PEPIL, my_QTL_thres, 1.5)
```

```{r my_tablevfdvdfbd4, results='asis', echo=FALSE}
print(xtable(summary_QTL_PEPIL), type = "html", include.rownames = F , comment=FALSE)
```

QTL 1B is present for BAR14 only, once more with the Silur allele beeing resistant. The 5A QTL is once more available, with the Dic2 allele beeing resistant.
Small QTL on 3B.
4B and 7B QTL are present as usual but nor for BLUPs!

We can have a look to the LOD-scores:
```{r}
# Variable to show on the plot?
my_var_to_show=c( Reduce(intersect, list(PEPIL,BLUP,AUDPC)), Reduce(intersect, list(PEPIL,BLUP,LAST))  )
#Chromosome 1B
plot_a_interactiveQTL(data, "1B", my_var_to_show, my_QTL_thres, 45)
#Chromosome 5A
plot_a_interactiveQTL(data, "5A", my_var_to_show, my_QTL_thres, 7)
```



## Type 2: Nb of Spikelet with Fusa
A lot of significant QTLs.

```{r, warning=FALSE}
summary_QTL_NEPIL=summary_table_for_QTL(data, NEPIL, my_QTL_thres, 1.5)
```

```{r my_tablevfdv4, results='asis', echo=FALSE}
print(xtable(summary_QTL_NEPIL), type = "html", include.rownames = F , comment=FALSE)
```

The 1B QTL is present for BAR14 as usual, with Silur as resistant. Warning, GRI13 as a QTL on the 1B also, but not on the same position.  
Small QTL on 1A, 2B, 3A, 3B, 4A, 6B, 7A.
QTL height and preco as usual, but not for BLUP.

We can have a look to the LOD-scores:
```{r}
# Variable to show on the plot?
my_var_to_show=c( Reduce(intersect, list(NEPIL,BLUP,AUDPC)), Reduce(intersect, list(NEPIL,BLUP,LAST))  )
par(mfrow=c(2,2))
#Chromosome 1B
plot_a_interactiveQTL(data, "1B", my_var_to_show, my_QTL_thres, 45)
#Chromosome 5A
plot_a_interactiveQTL(data, "5A", my_var_to_show, my_QTL_thres, 7)
```


[//]: =========================================================================================================================








[//]: =========================================================================================================================
#6/ Summary table

## Per experiment
Let's summarize what QTL has been found for each experiment. For each expe and each phenotype, we write QTL found, and the number of phenotypes that are significant for this QTL between bracklets.
```{r}
# initialize table
bil=data.frame(matrix(0,5,7))
colnames(bil)=c("CAP13", "GRI11", "GRI13", "GRI15", "LEC14", "BEQ11", "BAR14")
rownames(bil)=c("DON", "NOT","PEPI", "NEPIL", "PEPIL")
# fill it
# For each variable (DON, EPIL,...)
num=0
for(j in list(summary_QTL_DON, summary_QTL_NOT, summary_QTL_PEPI, summary_QTL_NEPIL, summary_QTL_PEPIL)){
  num=num+1
  
  # And for each experiment
  num2=0
  for(i in colnames(bil)){
     num2=num2+1
     a=j[ grep(i, j$carac ) , ]
     b=table(a$chromo)
     c=paste(names(b), " (", b, ")", sep="")
     d=paste(c, collapse = " - ")
     bil[num,num2]=d
  }}

# write when data is not available
bil[1,c(2,4,6,7)]="no data"
bil[2,c(1)]="no data"
bil[3,c(2)]="no data"
bil[5,c(2,6)]="no data"
```

```{r my_tableefzdde5, results='asis', echo=FALSE}
print(xtable(bil), type = "html", include.rownames = T , comment=FALSE)
```

## Details of all QTL
Let's save a huge table that will contain absolutely every QTL found as significant in this study. It will be a supplementary data for the publication.
```{r}
OR=rbind(summary_QTL_DON, summary_QTL_NOT, summary_QTL_PEPI, summary_QTL_NEPIL, summary_QTL_PEPIL)
```

## All signif QTL for publication
In the publication we will talk about QTL on blups only, when available, and on raw data if not. Let's list of these QTLs:
```{r}
OR=OR[which(OR$carac%in%PUBLI) , ]
OR=OR[order(c(OR$chromo, OR$carac)) , ]
OR$position=round(OR$position,2)
write.table(OR, file="~/Dropbox/Publi_Fusariose/SUPPORTING_DATA/OR_detail_all_QTLs.csv", row.names=F, col.names=T, quote=F, sep=";", dec=,)
```

## Save environment
```{r}
save(list=ls() , file="/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/QTL_R_environment.R")
```


[//]: =========================================================================================================================














