---
title: "Genetic Map Description"
output:
  rmdformats::material:
    highlight: kate
    self_contained: true
    code_folding: hide
    thumbnails: true
    gallery: true
    fig_width: 4
    fig_height: 4
---


#1/ Introduction

This file is a supplementary data attached with the publication concerning the genetic determinism of **Durum Wheat** to the **Fusarium head blight**. It aims to describe the **genetic map** provided in this paper. The script allowing to build the genetic map is in the folder **1_Partie_Bioinfo**.

Let's upload this genetic map:
```{r}
#Watch out, to reproduct analysis, you have to update the path.
map=read.table("/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/map_avec_posi_physique.txt" , header=T )

# number of markers
nmark=nrow(map)
```

This genetic map is composed of **`r nmark` markers**. Putative **physical positions** of markers are available.

Charge some **libraries** that will be useful
```{r results='hide', message=FALSE, warning=FALSE}
library(RColorBrewer)
library(xtable)
library(tidyverse)
library(rmdformats)
library(plotly)
library(knitr)
```







#2/ Map description  {.tabset .tabset-fade .tabset-pills}

## Summary table
Basic statistics are computed for **every chromosomes** of the genetic map, then for the **A** and **B genomes**, and finally for the **whole genetic map**. Results are presented in the table below:
```{r}
#Let's create a function that calculate some basic statistics for a piece of map
my_fun=function(my_map){
    num=nrow(bilan)
    num=num+1
    bilan[num,1]=i
    bilan[num,2]=nrow(my_map)
    bilan[num,3]=max(my_map[,3])
    gaps= sort(my_map[,3])[-1] - sort(my_map[,3])[-length(my_map[,3])]
    bilan[num,4]=mean(gaps)
    bilan[num,5]=max(gaps)
    bilan[num,6]=round(nrow(unique(my_map[,c(1,3)])),0)
    return(bilan)
    }

#Let's apply this function on our map, chromosome by chromosome and for the whole map:

# summary table that we are going to fill
bilan=data.frame(matrix(0,0,6)) ; num=0
colnames(bilan)=c("Chromo","nbr_marker","size_in_cM","average_gap","biggest_gap","nb_uniq_pos")

# apply the function to every chromosome
for(i in levels(map$group)){
    map_K=map[map$group==i,]
    bilan=my_fun(map_K)
    }
# then to A and B genomes
for(i in c("A" , "B")){
    map_K=map[substr(map$group , 2 , 2)==i , ]
    bilan=my_fun(map_K)
    }
# then to the whole map
i="tot"
bilan=my_fun(map)
```

```{r my_table1, results='asis', fig.align="center"}
#print(xtable(bilan), type = "html", include.rownames = F , comment=FALSE)
kable(xtable(bilan))

```

This table is saved as a **supplementary material** for the publication.
```{r}
write.table(bilan, "../../../SUPPORTING_DATA/OR_map_feature.csv", sep=";", quote=F, row.names = F)
```



## Markers distribution (plot)
```{r}
ggplot(map, aes(y=position, x=group, color=ref)) + 
  geom_point() +
  scale_y_reverse() 
```






#3/ Genetic vs Physical positions  {.tabset .tabset-fade .tabset-pills}
For most of the SNPs, a physical position is available trough the Ensembl database. It is interesting to compare the physical and the genetic positions of markers when both informations are available.



##Marey Map
This is the "marey map" representation. We will produce a figure for the paper with this code.

```{r, fig.align='center', fig.height=9, fig.width=7, warning=FALSE}
p=map %>% 
  filter(group==group_phy) %>% 
  ggplot(aes(x=position_phy, y=position, color=ref, text=marker)) + 
    geom_point(size=0.5, alpha=0.5) + 
    facet_wrap(~group, scales="free") + 
    theme(legend.position="none", axis.text=element_blank() , axis.ticks= element_blank() 
    )
ggplotly(p)
```

Save this figure as a supplementary material
```{r}
png("../../../SUPPORTING_DATA/OR_marey_map.png")
p
dev.off()
```



##Summary statistics

Let's calculate the spearman correlation between physical and genetic positions for each chromosome?
```{r}
tb=map %>%
  filter(group==group_phy & !is.na(group)) %>% 
	group_by(group) %>%
	summarize( 
	  cor_spearman=cor(position, position_phy, use="complete.obs", method="spearman") %>% round(2),
	  nvalue=length(position)
	  )
kable(xtable(tb))
```


##Recombination events
Do we have markers that have different chromosome attributions? Apparently we have some:  
- between 7A and 4A

```{r, fig.align='center', fig.height=9, fig.width=7, warning=FALSE}
adjacency=table(map$group, map$group_phy)
kable(xtable(adjacency))
diag(adjacency)=0
p=adjacency %>% 
  as.data.frame() %>% 
  rename(GeneticPosition=Var1, PhysicalPosition=Var2, NumberOfMarkers=Freq) %>% 
  ggplot( aes(x=GeneticPosition, y=PhysicalPosition, z=NumberOfMarkers, fill=NumberOfMarkers)) +
    geom_tile() + 
    theme_bw()
ggplotly(p)
```


	