---
title: "Genetic map analysis"
author: "Yan Holtz"
date: "30th May 2016"
output:
  rmdformats::readthedown:
    highlight: kate
  html_document:
    toc: yes
  pdf_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
  word_document: default
---


#1/ Introduction
This file is a supplementary data attached with the publication. It aims to describe the genetic map provided in this paper.

Let's upload this genetic map:
```{r}
#Watch out, to reproduct analysis, you have to update the path.
map=read.table("/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/map_avec_posi_physique.txt" , header=T )[,c(1:3)]
colnames(map)=c("chromo_map","marker","position_map")
```

Remember the number of markers in this genetic map?:
```{r}
nrow(map)
```

Let's upload the physical positions of genes:
```{r}
POS1=read.table("/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/Physical_map_of_EPO_on_BW.txt" , na.string="-" , header=T , dec=".") ; colnames(POS1)=c( "chromo_BW" ,  "contig" ,"position_BW" )
POS2=read.table("/Users/yan/Dropbox/Publi_Fusariose/ANALYSIS_REPRO/DATA/physical_map_of_BW.txt", na.string="-" , header=T , dec=".") ;  colnames(POS2)=c( "chromo_BW" ,  "contig" ,"position_BW" )
POS=rbind(POS1,POS2)
```

Charge some libraries that will be useful
```{r results='hide', message=FALSE, warning=FALSE}
library(RColorBrewer)
library(xtable)
```







#2/ Map description

## Summary table
Let's create a function that calculate some basic statistics for a piece of map
```{r}
my_fun=function(my_map){
    num=nrow(bilan)
    num=num+1
    bilan[num,1]=i
    bilan[num,2]=nrow(my_map)
    bilan[num,3]=max(my_map[,3])
    gaps= sort(my_map[,3])[-1] - sort(my_map[,3])[-length(my_map[,3])]
    bilan[num,4]=mean(gaps)
    bilan[num,5]=max(gaps)
    bilan[num,6]=round(nrow(unique(my_map[,c(1,3)])),0)
    return(bilan)
    }
```

Let's apply this function on our map, chromosome by chromosome and for the whole map:
```{r}
# summary table that we are going to fill
bilan=data.frame(matrix(0,0,6)) ; num=0
colnames(bilan)=c("Chromo","nbr_marker","size_in_cM","average_gap","biggest_gap","nb_uniq_pos")

# apply the function to every chromosome
for(i in levels(map$chromo_map)){
    map_K=map[map$chromo_map==i,]
    bilan=my_fun(map_K)
    }
# then to A and B genomes
for(i in c("A" , "B")){
    map_K=map[substr(map$chromo_map , 2 , 2)==i , ]
    bilan=my_fun(map_K)
    }
# then to the whole map
i="tot"
bilan=my_fun(map)
```

Let's have a look to these summary statistics. 
```{r my_table1, results='asis'}
print(xtable(bilan), type = "html", include.rownames = F , comment=FALSE)
```



## Markers distribution (plot)
```{r}
par(mar=c(0,0,0,0))
par(bg="grey")
nb_K=nlevels(map[,1])
num=0
plot(0,0 , xlim=c(1,nb_K) , ylim = rev(range(map[,3])) , bty="n" , xaxt="n" , yaxt="n" )
abline(h=seq(0 , max(map[,3]) , 50) , col="white" , lwd=0.2)
for (i in levels(map[,1])){
    num=num+1
    text(num-0.3 , 0 , i)
    a=map[map[,1]==i , ]
    points( rep(num,nrow(a)) , a[,3]  , pch=20 , cex=1, col=ifelse( regexpr("Traes", a$marker) != -1 | regexpr("TRAES",a$marker) != -1  , rgb(1,0,0,0.05) , rgb(0,0,1,0.1) )  )
    }
legend("bottomright", legend = c("BAITES" , "RNA_SEQ" ) , col = c(rgb(0.1,0.1,1,0.4) , rgb(1,0.1,0.1,0.4)) , bty = 1 , pch=20 , pt.cex = 2, cex = 0.8, horiz = FALSE, inset = c(0.1, 0.1))
```






#3/ Marey map
For most of the SNPs, a physical position is available trough the Ensembl database. It is interesting to compare the physical and the genetic positions of markers when both informations are available.


##Summary statistics

```{r}
# First we modify a little the files to be able to merge them. It gives the "data" object:
POS$chromo_BW=as.factor(substr(POS$chromo_BW , 1 , 2))
fun=function(x){strsplit(x,"@")[[1]][1] }
map$contig=unlist(lapply(as.character(map[,2]) , fun))
data=merge(map , POS , by.x=4 , by.y=2 , all=F)
data=data[ , -1]

# Compute the extremity of each chromosomes:
extremity=merge(aggregate(POS$position_BW , by=list(POS$chromo_BW) , min) , aggregate(POS$position_BW , by=list(POS$chromo_BW) , max) , by.x=1 , by.y=1)

# Function that computes statistics for a piece of map
my_fun=function(my_map , i){
  num=nrow(bilan2)
  num=num+1
  bilan2[num,1]=i
  bilan2[num,2]=nrow(my_map)
  bilan2[num,3]=max(my_map[,3])
  gaps= sort(my_map[,3])[-1] - sort(my_map[,3])[-length(my_map[,3])]
  bilan2[num,4]=mean(gaps)
  bilan2[num,5]=max(gaps)
  common_map=merge(map,my_map , by.x=4 , by.y=2 , all=F)
  common_map=common_map[as.character(common_map$chromo_map) == as.character(common_map$chromo_BW) , ]
  bilan2[num,6]=nrow(common_map)
  bilan2[num,7]=round(cor(common_map$position_map , common_map$position_BW , method="spearman"),2)
  return(bilan2)
}

# we will store the results in a "bilan2"" table:
bilan2=data.frame(matrix(0,0,7))
colnames(bilan2)=c("Chromo","nbr marker","size in cM","average gap","biggest gap", "nbr_common_marker" , "Spearman")

# compute statistics for each chromosome separately:
for(i in levels(map$chromo_map)){
  map_K=POS[POS$chromo_BW==i,]
  bilan2=my_fun(map_K,i)
}
# then for A and B genomes:
for(i in c("A" , "B")){
  map_K=POS[substr(POS$chromo_BW , 2 , 2)==i , ]
  bilan2=my_fun(map_K,i)
}
# then for the whole map:
i="tot"
bilan2=my_fun(POS , i)

#Add these informations to the summary of the genetic map:
bilan=merge(bilan, bilan2[,c(1,6,7)] , by.x=1 , by.y=1)
```

```{r my_table2, results='asis'}
print(xtable(bilan), type = "html", include.rownames = F , comment=FALSE)
```

It is the table number one of the publication, so we save it also:
```{r}
write.table(x=bilan,file="/Users/yan/Dropbox/Publi_Fusariose/TABLES/table1.csv", quote=F, col.names=NA , row.names=T, sep=";")
```


##Graphic Representation
This is the "marey map" representation. We will produce a figure for the paper with this code.
	
```{r, message=FALSE, warning=FALSE , fig.height=10}
# Plot preparation:
#png(filename = "/Users/yan/Dropbox/Publi_Fusariose/FIGURE/Fig_2.png" )
par(mfrow=c(4,4))
par(mar=c(2.5 , 2.5 , 0.5 , 0.5))

# We will store features of chromosomes in a summary table:
bilan_model=data.frame(matrix(0,14,6) ) ; colnames(bilan_model)=c("chromo" , "a" , "b x" , "c x2", "dx3" , "R2 adj")
bilan_outsider=data[0, ]
bilan_outsider2=data[0, ]
num=0

# For each chromosome:
for(i in levels(data$chromo_map)){
	
	# Plot each markers with both physical & genetic positions:
	don=data[data$chromo_map==i & data$chromo_BW==i & !is.na(data$chromo_BW) , ]
	if(nrow(don)==0){next}
	my_contigs=unlist(lapply(as.character(don$marker) , fun))
	plot(don$position_map ~ don$position_BW ,  xlab="" , ylab=""  , pch=20 , 
	     xlim=c(0,max(don$position_BW , na.rm=T)+20) , ylim=c(0,max(don$position_map , na.rm=T)+20) , 
	     col=ifelse( regexpr("Traes", don$marker) != -1 | regexpr("TRAES",don$marker) != -1  , "blue" , "red" ) , cex.axis=0.7)
	legend("topleft", paste(i," (",length(don$position_BW[!is.na(don$position_BW)]),")" , sep="") , 
	     cex=1 , text.col="black" , bty="n" )
	
	# Add 2 ablines for start and end of the chromosome
	abline(v=extremity[extremity[,1]==i ,c(2,3) ] , col="grey")
	
	# Let's add a polynomial regression:
	num=num+1
	bilan_model[num,1]=i
	y=don$position_map
	x=don$position_BW
	model=lm(y ~ x + I(x^2) + I(x^3))
	bilan_model[num , c(2:5)]=model$coefficients
	bilan_model[num , 6]=summary(model)$adj.r.squared
	myPredict <- predict( model , interval="predict" )
	ix <- sort(x,index.return=T)$ix
	lines(x[ix], myPredict[ix], col=2, lwd=2 )
	polygon(c(rev(x[ix]), x[ix]), c(rev(myPredict[ ix,3]), myPredict[ ix,2]), col = rgb(0.7,0.7,0.7,0.4) , border = NA)
	
	# Let's notice which are the SNPs far from the polynomial regression (="outsider")
	don$predict=myPredict[,1]
	AA=don[don$predict > don$position_map+15 | don$predict < don$position_map-15 , ]
	bilan_outsider=rbind(bilan_outsider , AA)
	
	# And notice which one are note in the confidence interval of the regression.
	confidence=    predict( model , interval="predict" )
	don$lwr=confidence[,2]
	don$upr=confidence[,3]
	BB=don[ don$position_map > confidence[,2] & don$position_map > confidence[,3] , ]
	bilan_outsider2=rbind(bilan_outsider2 , BB)
	
	}
```


	